README.rst
healthy
===========

healthy checks the health of a Python package from its pypi listing

.. image:: https://api.travis-ci.org/dustinmm80/healthy.png
    :target: https://travis-ci.org/dustinmm80/healthy

The package is given a score, based on this formula: ::

    Starting score                      = 0
    Has license                         + 20
    Has release files                   + 30
    Has homepage                        + 10
    Updated in last 6 months            + 15
    Has summary                         + 15
    Has description                     + 30
    Has author name or email            + 10
    Has Python classifiers              + 15

Results are color coded, based on the score

Usage
-----

The simplest command you can run is ::

    healthy Flask

This will check the latest version of Flask, and give you back its score.

To check a specific version of a package with verbose output ::

    healthy Flask 0.8 -v

There are 3 optional parameters ::

    package_version     : positional, to check the health of a specific version of a package
    -v, --verbose       : show reasons for the health scores the packages were given
    -n, --no-output     : print no output to the terminal, useful for integrating healthy with other packages


Sample Output
-------------

::

    healthy Django 1.2.1 -v

    Django v1.2.1
    -----
    score: 85/145 58%
    Description missing
    Python classifiers missing
    Package not updated in 180 days

Use Case
--------

``healthy`` can be used when you are trying to choose a package from a number of options. With ``healthy`` you can
quickly gauge the risk and benefit of one package over the other and make your decision easier.

Installation
------------
::

    pip install healthy

----

Supports Python 2.7, 3.2, 3.3

Python <=2.6, 3.0 and 3.1 are not supported, they don't have argparse


CHANGES.txt
v<0.1.0>, <May.12.2013> -- Initial release.
v<0.1.1>, <May.12.2013> -- Bugfix - don't return values unless using --no-output
v<0.1.2>, <May.22.2013> -- Scoring is done by percentage of possible points now; Windows terminal bugfix


tests/test_healthy.py
#!/usr/bin/env python
# coding=utf-8

"""
Tests for the checkmyreqs package
"""
import unittest
from healthy import calculate_health


class HealthyTestCases(unittest.TestCase):
    """
    Test cases for calculating the health of pypi packages
    """

    def test_package_no_version(self):
        """
        A user has passed in no version
        """
        health, reasons = calculate_health('Django', no_output=True)

        self.assertEqual(len(reasons), 0)

    def test_package_with_version(self):
        """
        A user has passed in a version
        """
        health, reasons = calculate_health('Django', '1.4.1', no_output=True)

        self.assertGreater(len(reasons), 0)

    def test_missing_package(self):
        """
        User passed in a package not on pypi
        """
        health, reasons = calculate_health('j8aj9d8whadjjslkh', '1.4.1', no_output=True)

        self.assertEqual(health, 0)

        self.assertEqual(len(reasons), 0)

if __name__ == '__main__':
    unittest.main()


healthy.py
#! /usr/bin/env python
# coding=utf-8

"""
healthy.py

Checks the health of a Python package, based on it's Pypi information
"""
import argparse
import os
import checks

try:
    # Different location in Python 3
    from xmlrpc.client import ServerProxy
except ImportError:
    from xmlrpclib import ServerProxy

try:
    from blessings import Terminal
except:
    class Terminal(object):
        """A dummy Terminal class if we can't import Terminal."""
        def passthrough(self, s):
            return s
        bold = red = green = yellow = orange = passthrough

TERMINAL = Terminal()

BASE_PATH = os.path.dirname(os.path.abspath(__file__))
CLIENT = ServerProxy('http://pypi.python.org/pypi')


def calculate_health(package_name, package_version=None, verbose=False, no_output=False):
    """
    Calculates the health of a package, based on several factors

    :param package_name: name of package on pypi.python.org
    :param package_version: version number of package to check, optional - defaults to latest version
    :param verbose: flag to print out reasons
    :param no_output: print no output
    :param lint: run pylint on the package

    :returns: (score: integer, reasons: list of reasons for score)
    :rtype: tuple
    """
    total_score = 0
    reasons = []

    package_releases = CLIENT.package_releases(package_name)
    if not package_releases:
        if not no_output:
            print(TERMINAL.red('{} is not listed on pypi'.format(package_name)))
        return 0, []

    if package_version is None:
        package_version = package_releases[0]

    package_info = CLIENT.release_data(package_name, package_version)
    release_urls = CLIENT.release_urls(package_name, package_version)

    if not package_info or not release_urls:
        if not no_output:
            print(TERMINAL.red('Version {} is not listed on pypi'.format(
                package_version)))
        return 0, []

    if not no_output:
        print(TERMINAL.bold('{} v{}'.format(package_name, package_version)))
        print('-----')

    checkers = [
        checks.check_license,
        checks.check_homepage,
        checks.check_summary,
        checks.check_description,
        checks.check_python_classifiers,
        checks.check_author_info,
        checks.check_release_files,
        checks.check_stale
    ]

    for checker in checkers:
        result, reason, score = checker(package_info, release_urls)
        if result:
            total_score += score
        else:
            reasons.append(reason)

    if total_score < 0:
        total_score = 0

    if not no_output:
        percentage = int(float(total_score) / float(checks.TOTAL_POSSIBLE) * 100)
        score_string = 'score: {}/{} {}%'.format(total_score, checks.TOTAL_POSSIBLE, percentage)

        print(get_health_color(percentage)(score_string))

    if verbose and not no_output:
        for reason in reasons:
            print(reason)

    if no_output:
        return total_score, reasons

# def lint_package(download_url):
#     """
#     Run pylint on the packages files
#     :param download_url: download url for the package
#     :return: score of the package
#     """
#     sandbox = create_sandbox()
#     package_dir = download_package_to_sandbox(sandbox, download_url)
#     pylint_score = score(package_dir)
#     destroy_sandbox(sandbox)
#
#     return pylint_score

def get_health_color(score):
    """
    Returns a color based on the health score
    :param score: integer from 0 - 100 representing health
    :return: string of color to apply in terminal
    """
    color = TERMINAL.green

    if score <= 80:
        color = TERMINAL.yellow
    elif score <= 60:
        color = TERMINAL.orange
    elif score <= 40:
        color = TERMINAL.red

    return color

def main():
    """
    Parses user input for a package name
    :return:
    """
    parser = argparse.ArgumentParser('Determines the health of a package')

    parser.add_argument(
        'package_name',
        help='Name of package listed on pypi.python.org',
    )

    parser.add_argument(
        'package_version', nargs='?',
        help='Version of package to check',
    )

    parser.add_argument(
        '-v', '--verbose', required=False,
        help='Show verbose output - the reasons for the package health score',
        action='store_true'
    )

    parser.add_argument(
        '-n', '--no_output', required=False,
        help='Show no output - no output will be generated',
        action='store_true'
    )

    args = parser.parse_args()

    return calculate_health(args.package_name, args.package_version, args.verbose, args.no_output)


if __name__ == '__main__':
    main()


package_utils.py
#! /usr/bin/env python
# coding=utf-8

"""
Utilities for fetching and unpacking packages from pypi
"""
import os
import shutil
import tarfile
import tempfile
import requests


def create_sandbox(name='healthybox'):
    """
    Create a temporary sandbox directory
    :param name: name of the directory to create
    :return: The directory created
    """
    sandbox = tempfile.mkdtemp(prefix=name)
    if not os.path.isdir(sandbox):
        os.mkdir(sandbox)

    return sandbox

def download_package_to_sandbox(sandbox, package_url):
    """
    Downloads an unzips a package to the sandbox
    :param sandbox: temporary directory name
    :param package_url: link to package download
    :returns: name of unzipped package directory
    """

    response = requests.get(package_url)

    package_tar = os.path.join(sandbox, 'package.tar.gz')

    with open(package_tar, 'w') as f:
        f.write(response.content)

    os.chdir(sandbox)

    with tarfile.open('package.tar.gz', 'r:gz') as tf:
        tf.extractall()

    directory = [d for d in os.listdir(sandbox) if os.path.isdir(d)][0]

    return os.path.join(sandbox, directory)


def destroy_sandbox(sandbox):
    """
    Destroys a temporary sandbox directory
    :param sandbox: name of the directory acting as sandbox
    """
    if os.path.isdir(sandbox):
        shutil.rmtree(sandbox)


def main():
    """
    Main function for this module
    """
    sandbox = create_sandbox()
    directory = download_package_to_sandbox(
        sandbox,
        'https://pypi.python.org/packages/source/c/checkmyreqs/checkmyreqs-0.1.6.tar.gz'
    )
    print(directory)
    destroy_sandbox(sandbox)


if __name__ == '__main__':
    main()


pylint_runner.py
#! /usr/bin/env python
# coding=utf-8

"""
Utilities for fetching and unpacking packages from pypi
"""
from collections import Counter
import fnmatch
import os

from pylint.lint import Run
from pylint.reporters.text import TextReporter

SCORING_VALUES = {
    'F': 5,
    'E': 4,
    'W': 3,
    'R': 2,
    'C': 1
}

def score(package_path):
    """
    Runs pylint on a package and returns a score
    Lower score is better

    :param package_path: path of the package to score
    :return: number of score
    """
    python_files = find_files(package_path, '*.py')

    total_counter = Counter()

    for python_file in python_files:
        output = run_pylint(python_file)
        counter = parse_pylint_output(output)
        total_counter += counter

    score_value = 0

    for count, stat in enumerate(total_counter):
        score_value += SCORING_VALUES[stat] * count

    return score_value / 5

def parse_pylint_output(output):
    """
    Parses pylint output, counting number of errors, conventions, etc
    :param output: output list generated by run_pylint()
    :return:
    """

    stripped_output = [x[0] for x in output]

    counter = Counter(stripped_output)

    return counter

class WritableObject(object):
    """
    Dummy output stream for pylint"
    """
    def __init__(self):
        self.content = []
    def write(self, st):
        "dummy write"
        self.content.append(st)
    def read(self):
        "dummy read"
        return self.content

def run_pylint(filename):
    """
    Runs pylint on a given file
    :param filename:
    :return: list of pylint errors
    """
    ARGS = [
        '-r',
        'n'
    ]
    pylint_output = WritableObject()
    Run([filename]+ARGS, reporter=TextReporter(pylint_output), exit=False)

    lines = []
    for line in pylint_output.read():
        if not line.startswith('*') and line != '\n':
            lines.append(line)

    return lines

def find_files(directory, pattern):
    """
    Recusively finds files in a directory re
    :param directory: base directory
    :param pattern: wildcard pattern
    :return: generator with filenames matching pattern
    """
    for root, __, files in os.walk(directory):
        for basename in files:
            if fnmatch.fnmatch(basename, pattern):
                filename = os.path.join(root, basename)
                yield filename


checks.py
#! /usr/bin/env python
# coding=utf-8

"""
checks.py

Check functions that healthy employs to generate a score
"""
from datetime import datetime, timedelta

DAYS_STALE = 180  # number of days without update that a package is considered 'stale'

# Points
HAS_LICENSE             = 20
HAS_RELEASE_FILES       = 30
NOT_STALE               = 15
HAS_SUMMARY             = 15
HAS_DESCRIPTION         = 30
HAS_PYTHON_CLASSIFIERS  = 15
HAS_AUTHOR_INFO         = 10
HAS_HOMEPAGE            = 10

TOTAL_POSSIBLE = sum(
    [
        HAS_LICENSE, HAS_RELEASE_FILES, NOT_STALE, HAS_SUMMARY, HAS_DESCRIPTION, HAS_PYTHON_CLASSIFIERS,
        HAS_AUTHOR_INFO, HAS_HOMEPAGE
    ]
)

BAD_VALUES = ['UNKNOWN', '', None]


def check_license(package_info, *args):
    """
    Does the package have a license classifier?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None, score to be applied)
    """
    classifiers = package_info.get('classifiers')
    reason = "No License"
    result = False

    if len([c for c in classifiers if c.startswith('License ::')]) > 0:
        result = True

    return result, reason, HAS_LICENSE


def check_homepage(package_info, *args):
    """
    Does the package have a homepage listed?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None)
    """
    reason = "Home page missing"
    result = False

    if package_info.get('home_page') not in BAD_VALUES:
        result = True

    return result, reason, HAS_HOMEPAGE


def check_summary(package_info, *args):
    """
    Does the package have a summary listed?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None)
    """
    reason = "Summary missing"
    result = False

    if package_info.get('summary') not in BAD_VALUES:
        result = True

    return result, reason, HAS_SUMMARY


def check_description(package_info, *args):
    """
    Does the package have a description listed?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None)
    """
    reason = "Description missing"
    result = False

    if package_info.get('description') not in BAD_VALUES:
        result = True

    return result, reason, HAS_DESCRIPTION


def check_python_classifiers(package_info, *args):
    """
    Does the package have Python classifiers?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None, score to be applied)
    """
    classifiers = package_info.get('classifiers')
    reason = "Python classifiers missing"
    result = False

    if len([c for c in classifiers if c.startswith('Programming Language :: Python ::')]) > 0:
        result = True

    return result, reason, HAS_PYTHON_CLASSIFIERS


def check_author_info(package_info, *args):
    """
    Does the package have author information listed?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None)
    """
    reason = "Author name or email missing"
    result = False

    if package_info.get('author') not in BAD_VALUES or package_info.get('author_email') not in BAD_VALUES:
        result = True

    return result, reason, HAS_AUTHOR_INFO


def check_release_files(package_info, *args):
    """
    Does the package have release files?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None)
    """
    reason = "No release files uploaded"
    result = False

    release_urls = args[0]
    if len(release_urls) > 0:
        result = True

    return result, reason, HAS_RELEASE_FILES


def check_stale(package_info, *args):
    """
    Is the package stale?
    :param package_info: package_info dictionary
    :return: Tuple (is the condition True or False?, reason if it is False else None)
    """
    reason = 'Package not updated in {} days'.format(DAYS_STALE)
    result = False

    now = datetime.utcnow()

    release_urls = args[0]
    if len(release_urls) > 0:
        package_uploaded_time = release_urls[0]['upload_time']
        if now - timedelta(days=DAYS_STALE) <= package_uploaded_time:
            result = True

    return result, reason, NOT_STALE
